### ISSUE TITLE
executeCommand() doesn't execute the checkpoint_paste() command

### ISSUE THREAD

**Issue by mitkonikov** [Manim-Notebook/manim-notebook#18]
Terminals: GitBash and Conda through Powershell and GitBash
Issue replicated on 2 different PCs both Windows 11 one Enterprise and one Education (Version 23H2)
IPython version: 8.21.0

I've downloaded the repository, npm installed, npm compiled everything and attached the debugger.

Simple Manim Scene:

```py
from manimlib import *

class SubTensors(InteractiveScene):
    def construct(self):
        ## Interactions
        circle = Circle(radius = 2)
        self.play(DrawBorderThenFill(circle))
```

Detects the preview cell.
Run the command:

```sh
manimgl sub_tensor.py SubTensors -se 5
```

And when I click on the Preview Manim Cell, it clears the console, writes checkpoint_paste() but doesn't run it at all.

With the debugger, I can see that with both the `executeCommand` and `terminal.sendText`, the command is not executed.
If then I manually click ENTER in terminal, the command is executed, however it's now too late, the clipboard is already reset to the previous state and throws an error because it tries to run something that is not python code.

Since this happens with different terminals, I believe it's something due to the IPython console. Because if we just click `Preview Manim Cell`, without running the `manimgl` command, the command `checkpoint_paste()` is executed properly (even though, in this case, it's an error, since it's not recognized as a Powershell/Bash/Conda command).

What I have already tried:
 - Adding `\r\n` or `\n` at the end of the `PREVIEW_COMMAND` string
 - Adding spaces or other chars at the end of the `PREVIEW_COMMAND` string
 - `terminal.sendText` actually has a second parameter that is used to execute the command, setting that to `true`
 - Executing always `sendText` regardless if the terminal supports `shellIntegration`

Nothing worked. Any ideas as to why would this happen? I'm not familiar with the IPython console; it may have some weird interaction behaviour.

**Comment by VladimirFokow** [comment]
Hello!

@Splines , for info: the relevant [code section](https://github.com/bhoov/vscode-manim/blob/f0b23914b24b53bfdbaeee5ff746b11e09b9ed85/src/previewCode.ts#L52-L54)

So the problem is:
`checkpoint_paste()` is written,
but is NOT executed.

<br />

@mitkonikov I really don't know the reason... Could you record a video?

Here are some thoughts:

I don't think the behavior of the IPython console is special. I mean it works for us..
Maybe it's the `terminal.shellIntegration` who is broken, and it just doesn't reload the VSCode extension properly after changing something for experimentation?
Really, even always just `send_text` - doesn't work?

**Comment by bhoov** [comment]
I would also appreciate a video to understand. Right now, I have a suspicion that you are starting the `manimgl ... -se ...` command from a terminal other than the integrated VSCode terminal. Right now, our extension sends the `checkpoint_paste()` command only to the integrated, active VSCode terminal (i don't think we have tested multiple terminal windows open in the same vscode instance)

Can you try running the command in the integrated VSCode terminal and see if things work as expected?

**Comment by mitkonikov** [comment]
Answers to the previous questions:
 - I'm running the `manimgl` from the integrated terminal of course
 - Even if I restart the extension it doesn't execute it
 - `even if always just send_text - it doesn't execute in the IPython shell, but executes in the bash shell? That's weird` - this is true, it executes it everywhere, except the IPython shell. But notice that the IPython shell could be doing some string trimming or smth else itself. That is why my suspicion is in the IPython shell.

Here's a video I made going through everything, including the preparation, debugger, different terminals. I forgot to show you that it works if we are executing the command on a terminal without running the `manimgl` command at all. Sorry for the lengthy video, I just wanted to show you everything how it behaves.

https://youtu.be/ahBg4jx75Ds

**Comment by mitkonikov** [comment]
Here's an image. I wanted to see if it would execute the checkpoint_paste() even inside `conda` terminal/env, without running `manimgl`. It executes it immediately.

![image](https://github.com/user-attachments/assets/2b47ffae-bc93-470c-85e5-094832500312)

**Comment by mitkonikov** [comment]
OMG. I fixed it.. I think..

```js
const PREVIEW_COMMAND = `\x0C checkpoint_paste()\x1b`; // \x0C is Ctrl + L
```

When VSCode is writing text in the command line, the cursor is not being updated to the end. So, IPython shell thinks that the cursor is in the beginning, thus not executing the command but rather adding a new line. This behaviour is explained in this [issue](https://github.com/ipython/ipython/issues/10425). So, after adding an ESC symbol at the end of the `PREVIEW_COMMAND`, IPython then sees that `\n` character should execute the command.

**Comment by VladimirFokow** [comment]
omg, wow

**Comment by VladimirFokow** [comment]
for reference, here is the [3b1b's command](https://github.com/3b1b/videos/blob/4203c7a9a54842b98c943d7d8f5d85dea330c543/sublime_custom_commands/manim_plugins.py#L65-L70) - there I didn't know what the `"\x7F" * col if clear else "",  # Bad hack` was.. Maybe for this

**Comment by mitkonikov** [comment]
According to [this](https://en.wikipedia.org/wiki/Delete_character), 7F is the Delete character. This could be the same fix. Because the `Delete` character is a special one, it could just reset the cursor to the end. Let me try it.

However, tell me if this works on your machines too. You can fix it or tell me to make a PR, it's just one line change.

**Comment by VladimirFokow** [comment]
Yes, I'll test and can push, or you can make a PR if you want to. 
I feel your solution is better than with x7F

**Comment by mitkonikov** [comment]
For me, 7F doesn't work at the beginning or at the end of the `PREVIEW_COMMAND`.

If I put `\x7f` at the beginning of the `PREVIEW_COMMAND` nothing changes (same command, no execution), however at the end, for some reason, it deletes the last character. I suspect that the cursor is not at the end, but one character behind. That could be some VSCode issue, not updating the cursor for the last character properly. I'm not sure though.

This is after executing: `\x0C checkpoint_paste()\x7f`
![image](https://github.com/user-attachments/assets/44b7bdc4-889d-4a73-bde6-cfc533bee78e)

**Comment by VladimirFokow** [comment]
your solution works for me 🚀

**Comment by RickLuiken** [comment]
Hi, I found this issue as well. I just tried the proposed solution by @mitkonikov. It seems to work most of the time, however when the window get a bit laggier (when you have a few updaters running), it stops working.

As for the \x7f in 3b1b's code, I guess it's there to clear whatever is in the terminal already. You see him getting the amount of chars in the terminal window, and adding \x7f that amount of times to clear the screen

**Comment by mitkonikov** [comment]
@RickLuiken Can you send some sample code, so I can try?

**Comment by mitkonikov** [comment]
> As for the \x7f in 3b1b's code, I guess it's there to clear whatever is in the terminal already. You see him getting the amount of chars in the terminal window, and adding \x7f that amount of times to clear the screen

This is true. Maybe this functionally is going to be needed in this extension as well. If there has been some error, that made it so the IPython shell has stopped executing. The characters will maybe remain in the terminal window? This has to be tested as well.

**Comment by mitkonikov** [comment]
> Hi, I found this issue as well. I just tried the proposed solution by @mitkonikov. It seems to work most of the time, however when the window get a bit laggier (when you have a few updaters running), it stops working.

I replicated this by just waiting for 4 seconds.

This is my code:
```py
from manimlib import *

class SubTensors(InteractiveScene):
    def construct(self):
        ## Interactions
        circle = Circle(radius = 2)
        self.play(DrawBorderThenFill(circle))
        self.wait(4)

        ## Remove
        self.remove(circle)
```

When I execute the first block, the wait for 4 seconds, makes the terminal busy, so if I don't wait until the end, and preview the first block again, the command will be written in the shell, but the ESC key is overlooked somehow. This is why I think, in case of executing, if the user clicks the preview again, it should first send a CTRL+C or something similar in order to stop executing and afterwards send the preview command. Let's try injecting CTRL+C in the PREVIEW_COMMAND..?

**Comment by VladimirFokow** [comment]
reopening, in case @RickLuiken  continues having issues..

**Comment by RickLuiken** [comment]
```python
from manimlib import *

class LaggyScene(Scene):

    def construct(self):
        ## create a bunch of boxes
        grid = Square().get_grid(100, 100, buff=0)
        grid.set_height(5)
        grid.set_stroke(WHITE)
        self.play(ShowCreation(grid))

        ## move the boxes using updaters
        position = ValueTracker(0)

        grid.add_updater(lambda m, position=position: m.move_to((position.get_value(), 0, 0)))

        self.play(position.animate.set_value(2), run_time=5)

        ## moves boxes back
        self.play(position.animate.set_value(0), run_time=5)
```

Here's a minimal example. I start running with `manimgl laggyscene.py LaggyScene -se 10`. Then run the second cell using the preview command. This one works fine with @mitkonikov's fix. After this animation, manim becomes a bit laggy. Running the third cell using the preview command then fails.

**Comment by mitkonikov** [comment]
Wow, I can replicate it as well. Even waiting for all of the commands to finish, the third block's command is not executed. Even with my fix.. :(

**Comment by mitkonikov** [comment]
> This is why I think, in case of executing, if the user clicks the preview again, it should first send a CTRL+C

Sending just `CTRL+C` as `\x03`, it stops the execution, but then it doesn't execute the new `checkpoint_paste()` command.
You have to click once again, to execute the new checkpoint paste command.

It's so weird.

**Comment by RickLuiken** [comment]
One way of fixing it is delaying the newline by doing:
```typescript
terminal.sendText(PREVIEW_COMMAND, false);
setTimeout(() => terminal.sendText(""), 100);
```
This is also done in this [ipython vscode extension](https://github.com/hoangKnLai/vscode-ipython/blob/c11002a968ca9bffe50ceac69cacca3476bdf699/src/ipython.ts#L383). However this adds the problem that we don't know how long the delay should be. 100 ms seems to solve my issues, but maybe even laggier scenes require more delay?

**Comment by mitkonikov** [comment]
This to me, suggests that the cursor is not being updated to the last position in time. But, as I can see, the VSCode's Terminal API doesn't allow for anything like "cursor position change" or waiting for something like this. I don't believe that the delay is even in ms, it should be very fast. Can you try pushing it to 1-10ms? I will try this on my machine and report the findings.

**Comment by RickLuiken** [comment]
On my slower laptop, 50ms seems to work sometimes. I've not seen 100ms fail. Adding a setting to change the delay and setting the default to 100ms seems okay to me!

**Comment by bhoov** [comment]
Question @RickLuiken, are you also working on windows? I cannot replicate this bug on MacOS since sending Ctrl+C to the terminal kills the entire interactive manim session (see #16, arguably a bigger bug than an enter command not being executed)

**Comment by VladimirFokow** [comment]
I tried laggyscene on MacOS - it was not laggy for me.. But that's just on my machine

It would be beneficial to solve this bug for all machines

**Comment by bhoov** [comment]
I also am unable to reproduce this bug on MacOS. My understanding is that @VladimirFokow and I are on MacOS, and we are unable to send Ctrl+C to the manimgl terminal when an existing command is running at all -- if we do, our entire session quits and we get the following warning which is followed by a silent exiting of the entire interactive session:

```
In [1]:  checkpoint_paste()
[11:36:55] INFO     Using the default configuration file, which you can modify in                 config.py:361
                    `/Users/hoo/miniconda3/lib/python3.10/site-packages/manimlib/default_config.y              
                    ml`                                                                                        
           INFO     If you want to create a local configuration file, you can create a file named config.py:362
                    `custom_config.yml`, or run `manimgl --config`                                             
0 ShowCreationVGroup:   3%| 1  /30   1.05s/it^C%                                                               /Users/hoo/miniconda3/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

I have continued to debug this MacOS-specific issue around Ctrl+C and it looks like the problem is in the IPython's [`InteractiveShellEmbed`](https://github.com/ipython/ipython/blob/6aa7cdc7d5f80edcd083865919e78cd3419e88e7/IPython/terminal/embed.py#L119) (see my comment [here](https://github.com/bhoov/manim-notebook/issues/16#issuecomment-2438161293)). Until this is fixed, we MacOS users cannot debug this issue of interrupting a laggy animation and having subsequent `checkpoint_paste()` commands not execute.

**Comment by Splines** [comment]
Even though this issue is just one week old, the codebase changed a lot in the meantime since this repo is in its initial building phase where things might break often.

Could you please check if this is still an issue for you on `main`? If yes, checkout the branch `fix/exit-scene` from PR #48, which has one more bug fix in it. If it is still is an issue, please send us logs by following [these instructions](https://github.com/bhoov/manim-notebook#troubleshooting).

For me, having checked out the code from `fix/exit-scene`, everything works as expected (or maybe I miss the point of this PR, in which case, please summarize it for me ;))

https://github.com/user-attachments/assets/94fa682c-1936-44f7-b2cc-24da80d233dd

**Comment by VladimirFokow** [comment]
@Splines , should you simplify your instructions above? (only on main branch)

**Comment by mitkonikov** [comment]
Unfortunately, I'm on a vacation at the moment with no access to my PC even remotely. I will be back on the 6th of November for sure. Until then, I can't confirm nor deny the existence of this issue. Please have it open until then. 🙏 

To recreate the problems we had, either run a scene with `wait(4)` in it (see example in some comment above) where you will cancel it while waiting and immediately run the preview again. **This should write the checkpoint paste command but not execute it.** Or there was this Laggy Scene (quoted below) where the same thing would sometimes happen after running the 3rd block. I then got this to happen on some other scenes I was working on, so it wasn't a scene-specific thing.


> ```python
> from manimlib import *
> 
> class LaggyScene(Scene):
> 
>     def construct(self):
>         ## create a bunch of boxes
>         grid = Square().get_grid(100, 100, buff=0)
>         grid.set_height(5)
>         grid.set_stroke(WHITE)
>         self.play(ShowCreation(grid))
> 
>         ## move the boxes using updaters
>         position = ValueTracker(0)
> 
>         grid.add_updater(lambda m, position=position: m.move_to((position.get_value(), 0, 0)))
> 
>         self.play(position.animate.set_value(2), run_time=5)
> 
>         ## moves boxes back
>         self.play(position.animate.set_value(0), run_time=5)
> ```
> 
> Here's a minimal example. I start running with `manimgl laggyscene.py LaggyScene -se 10`. Then run the second cell using the preview command. This one works fine with @mitkonikov's fix. After this animation, manim becomes a bit laggy. Running the third cell using the preview command then fails.

I'm replying to this from my phone, so if I missed something, please write back, I'll answer as quickly as I can. Thank you.

**Comment by RickLuiken** [comment]
Hi @Splines!

I just tested it with the lastest commit, the bug seems to still be happening. I have recorded a video to show what happens.

https://github.com/user-attachments/assets/8cf429ac-1bc0-47fd-9959-547fca423d00

**Comment by Splines** [comment]
Hey, thanks for the updates. I can unfortunately not reproduce this on Linux, so I'd need some logs from you. Please follow the steps of the [Troubleshooting section](https://github.com/Manim-Notebook/manim-notebook?tab=readme-ov-file#troubleshooting).

**Comment by RickLuiken** [comment]
Here you go: [Manim Notebook.log](https://github.com/user-attachments/files/17607997/Manim.Notebook.log)

**Comment by Splines** [comment]
@RickLuiken Thank you. However, I just noticed that this log file is not helping me out in this special case. Could you please do a

```
git checkout b19212b86795223
```

Then, do the same steps and record the log file? This time, it will also preserve the ANSI escape sequences that will probably play a crucial role here.

**Comment by RickLuiken** [comment]
@Splines here's the second log: [Manim Notebook.log](https://github.com/user-attachments/files/17610210/Manim.Notebook.log)

I looked through it a bit, but I didn't spot any obvious problem. One solution instead of the delay could be to read the terminal output and detect when this problem occurs.

**Comment by Splines** [comment]
From [this comment](https://github.com/Manim-Notebook/manim-notebook/issues/18#issuecomment-2431146809):

> So, after adding an ESC symbol at the end of the `PREVIEW_COMMAND`, IPython then sees that `\n` character should execute the command.

However, notice that the `\x1b` is just the escape code to _start_  an ANSI control sequences. See [this website](https://notes.burke.libbey.me/ansi-escape-codes/):

> ANSI escapes always start with `\x1b`, or `\e`, or `\033`.

> The majority of these escape codes start with `\x1b[`. This pair of bytes is referred to as `CSI`, or “Control Sequence Introducer”. By and large, a control sequence looks like:

```
0x1B + "[" + <zero or more numbers, separated by ";"> + <a letter>
```

That's why I'm not too sure if the change introduced in #23 is not actually more harm than help. E.g. subsequent commands in the shell could be misinterpreted because we inserted this `\x1b`.

---

Could you maybe play around a bit with this line:

https://github.com/Manim-Notebook/manim-notebook/blob/9e1b6f42b5573d445cf284ea5a72a4d4819e9823/src/previewCode.ts#L7

Try out different things and maybe start with the bare-bone `checkpoint_paste()` command without any control sequences. `\x0C` was just added for convenience (it corresponds to `Ctrl+L`). So, e.g. try out these commands (don't forget to reload the extension window after each one):

- `checkpoint_paste()`
- `checkpoint_paste()\n`
- ...

---

Maybe also related: https://github.com/ipython/ipython/pull/10489 (`Prefer execution when there's only a single line entered`). However that was long ago and I'm not sure if that is still used in their code today.

**Comment by Splines** [comment]
Thanks @RickLuiken for the fix in #81, I hope it works for the other Windows users in this issue as well ;)

**Comment by Splines** [comment]
Hopefully fixed again via #110 by appending `ESC + ENTER` at the end to force IPython into _not_ going into multi-line edit mode and instead actually executing the command.

**Comment by Splines** [comment]
This is still an issue, see #113

**Comment by Splines** [comment]
In my comment [here](https://github.com/ipython/ipython/issues/13054#issuecomment-2600235472) I link to all related issues. This is probably an issue in python-prompt-toolkit (https://github.com/prompt-toolkit/python-prompt-toolkit/issues/1894) that IPython consumes (https://github.com/ipython/ipython/issues/13054).

> [!tip]
> [This comment](https://github.com/kiblee/tod0/issues/62#issuecomment-2502205034) is a nice and brief summary of why this bug occurs including a possible workaround.

**Comment by Splines** [comment]
As described, this issue (https://github.com/prompt-toolkit/python-prompt-toolkit/issues/1894) might be our problem. To verify this, could you @mitkonikov @RickLuiken try to patch it locally and manually on your Windows machine, please?

- That is, navigate to the `site-packages` folder of your Python installation (easy if you've used `venv` to install manimlib). In that folder, go to `prompt_toolkit/input/win32.py`. In that file, search for the line `self.recognize_paste = recognize_paste` and replace it by `self.recognize_paste = False`.
- Maybe add a print statement there as well to make sure you've patched the right file for the right Python version.
- Then, try to preview the laggy scene again, but not on the branch of #113, since I made all kinds of other modifications there that could bias the result). Does it work now?

If yes, could you verify if the following patch works? It doesn't work in the pipeline unfortunately, so I'm not sure that it does what it is supposed to do. It does print `... has been patched` all the time in the GitHub Actions runs, but we still experience the bug. To apply the patch, prepend the following Python code at the top of your Python file (even before the manimlib import):

```py
import sys


def monkey_patch_win32_console_input_reader():
    if sys.platform != "win32":
        return

    try:
        print("Trying to patch prompt_toolkit.input.win32 ...")
        import prompt_toolkit.input.win32

        win32 = sys.modules["prompt_toolkit.input.win32"]
        original_class = win32.ConsoleInputReader

        # Define a patched version of the class
        class PatchedConsoleInputReader(original_class):
            def __init__(self, *args, **kwargs):
                if "recognize_paste" in kwargs:
                    kwargs["recognize_paste"] = False
                super().__init__(*args, **kwargs)

        win32.ConsoleInputReader = PatchedConsoleInputReader  # type: ignore

        print("prompt_toolkit.input.win32 has been patched.")
    except ImportError as exc:
        raise ImportError("Module 'prompt_toolkit.input.win32' could not be imported.") from exc


monkey_patch_win32_console_input_reader()
```

**Comment by mitkonikov** [comment]
@Splines on which commit should I be?

**Comment by Splines** [comment]
@mitkonikov Just try the current `main` branch.

**Comment by mitkonikov** [comment]
On the main branch, I get the following error:

```
Your ManimGL version could not be determined.
```

It might be due to this:

```
$ C:\GitHub\animations\.venv\Scripts\python.exe\Scripts\manimgl --version
bash: C:GitHubanimations.venvScriptspython.exeScriptsmanimgl: command not found
```

It assumes these are escape symbols instead of `\`. I'll patch this for now locally. But I'm reporting it, so you know what's going on.

**Comment by mitkonikov** [comment]
Also:

```
C:\GitHub\animations\.venv\Scripts\python.exe\Scripts\manimgl: No such file or directory
```

Look at the directory names...

**Comment by Splines** [comment]
@mitkonikov Thanks, would appreciate if you could open a new issue for these windows path problems. I can't actually test this since in the pipeline, we stub the path to the `manimgl` binary.

**Comment by Splines** [comment]
@mitkonikov Ah, I forgot that the main branch has the current fix applied, so to test it properly, you should also comment out these lines:

https://github.com/Manim-Notebook/manim-notebook/blob/a67282e1c9f09a34bfb2612b1552d860c72896c6/src/manimShell.ts#L778-L784

**Comment by mitkonikov** [comment]
## Preliminary testing:

 - Windows 11 Enterprise
 - vscode-manim commit (main at the time of speaking): 210d48c918b29e4e867e355cb60d32b9c38a8a76 
 - manim: pypi release 1.7.2 (fresh install)
 - default terminal: bash

For me, scene laggy with 16x16 is not laggy at all, so I had to make the grid 100x100. 

#### This was me talking about the main branch with the fix applied:
~~If I let the animations play out until the end, and I preview them one by one and get back to the animations, I cannot recreate the issue. The only time the command is not executed is if an animation is playing and I click to preview the next block. Consistently, the `checkpoint_paste` call will be copied, but not executed. However, on not-laggy scenes, I couldn't get even this to happen. More testing is also required. Is this the same problem we are encountering in the GitHub actions, or are you letting it play out until the end, before executing the next command?~~

#### This is me recreating the issue on the main branch with the fix commented out:
So, for the base case, I went and executed the laggy scene blocks. Executing the first, second, were OK, and when tried to execute the third block the issue occurred. I went and patched `recognize_paste` to be False always in `win32.py` directly. And it seems to work. There are no issues replaying the blocks. Even spamming the Preview Manim Cells on different blocks without letting them finish seems to always work correctly. 

The next thing was to try to apply the monkey_patch directly from the code. Don't forget to revert the `win32.py` to the original state.  In my testing, before and after the `import manimlib` command, the external patch doesn't work. With the internal patch (in `win32.py`) it works as it supposed to (as far as this preliminary testing goes).

**Comment by Splines** [comment]
@mitkonikov Yeah, thanks for your findings, this might confirm that it is indeed the annoying feature of the prompt-toolkit that IPython consumes which is responsible for the bug. https://github.com/prompt-toolkit/python-prompt-toolkit/issues/1894

Only thing to do now is to get my [monkey patch](https://github.com/Manim-Notebook/manim-notebook/issues/18#issuecomment-2600831075) to work ;) But at least, I know now that it could work in the pipeline if my monkey patch was correct. It's probably an issue in how modules are reloaded by manim itself. The reloading overrides our patched class parameter again. I will look into it.

**Comment by Splines** [comment]
@mitkonikov @RickLuiken It seems that indeed, this is the underlying issue for our problem: https://github.com/prompt-toolkit/python-prompt-toolkit/issues/1894. Now, I provide a new monkey patch that works in the pipeline 🎉 (I ran the test through twice). Please verify it is also working locally for you on a Windows machine.

- Make sure any previous manual modifications to `prompt_toolkit/input/win32.py` are reverted
- `git fetch`, `git checkout 5850e61` (this is a commit on #117, where the [CI/CD run](https://github.com/Manim-Notebook/manim-notebook/actions/runs/12854682540/job/35839465961) succeeded). In this commit, all previous fixes for this issue are removed and *only* my new monkey patch fix is applied.
- Run the extension, copy the contents from `tests/fixtures/laggy-patched.py` into a new file on the VSCode window that runs the extension and try to preview Manim Cells in any order. Try to break it.
- If this was successful, try to also run the tests locally. See the [dev guide](https://github.com/Manim-Notebook/manim-notebook/wiki/Developing).

Should this be successful, I will continue to work on #117 to provide this patch automatically for Windows users.

**Comment by mitkonikov** [comment]
@Splines 

If I understood correctly, I copied the contents `laggy-patched.py` and tested them. This way it works perfectly, as it is supposed to. But, without the patch (only the scene), even running the first block, doesn't execute the `checkpoint_paste` command. So, I guess the patch really works. Good job! 💯 

I have some issues, however, running the tests locally because some of the commands are Linux-based and don't work on Windows. I'll report them in separate issues and get back to you when I have a bit more time to test and fix stuff. This week, I have 3 exams and 2 presentations, so I'm on a really tight schedule.

**Comment by Splines** [comment]
> This way it works perfectly, as it is supposed to.

Nice, I'm happy that I could finally identify the bug ;) Thanks for testing. I will try to come up with a solution for how to auto-inject this patch such that the end-user will never see it. This won't be too easy since we also have to take care of the new Manim auto-reload mechanism (if I wanted to patch Manim itself). But I'm certain there is a solution.

> I have some issues, however, running the tests locally because some of the commands are Linux-based and don't work on Windows.

Curious to know what kind of commands this encompasses since I don't remember using anything Linux-specific. And for paths I hope I've always used `path.join()`. But I will see in the new issues you create.

> and get back to you when I have a bit more time to test and fix stuff. This week, I have 3 exams and 2 presentations, so I'm on a really tight schedule.

Wow, that's really a tight schedule, good luck with everything and take it easy.
